### PR TITLE
added multi-stage dockerfile and change main.go host port to 0.0.0.0

### PR DESCRIPTION
**Pull request by ZachSellman**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+RUN mkdir /app
+WORKDIR /app
+COPY . .
+RUN go mod download
+RUN CGO_ENABLED=0 GOOS=linux go build -o main ./...
+
+
+FROM alpine:latest AS production
+COPY --from=build-stage /app .
+EXPOSE 8080
```
Comment (justincely) [inline Dockerfile:13]: does actually doesn't do anything useful.  you can remove it.  EXPOSE is kinda a hint for certain use cases of docker, but it doesn't actually bind the port.

File: main.go [hunk 2]
```diff
@@ -8,7 +8,7 @@ import (
 )
 
 // This will be loaded from env later I'm assuming. For now just returns what port is used on localhost.
-const HOST = "127.0.0.1"
+const HOST = "0.0.0.0"
```
Comment (justincely) [inline main.go:11]: why this change?

**Review by justincely — CHANGES_REQUESTED** [review]

File: main.go (hunk 2 above)
Comment (ZachSellman) [reply main.go:11]: Was going to ask you about this. To externally test if the server is running properly I have to use 0.0.0.0 to access the container via it's IP and the port. Alternatively I can exec into the container, install curl and confirm that it's working by curling localhost:8080.

What I'm having a hard time understanding is which we are looking for at this point? If we're going to direct traffic to it eventually wouldn't it need to be 0.0.0.0? Or does it not matter since a K8 service will handle that connection later?

File: Dockerfile (hunk 1 above)
Comment (ZachSellman) [reply Dockerfile:13]: You got a typo at the start there so I'm not sure what you mean lol. Are you saying the whole production portion isn't needed or just the EXPOSE?

Comment (justincely) [reply Dockerfile:13]: Hah sorry, the `EXPOSE` bit doesn't actually do anything.  the `production` part is great - it's literally just an annotation to the docker engine on what TO expose if you use certain commands.  but with a deployment in k8s it has no effect.  And since we're going to make the port configurable, it's good to remove and values like this that remain static and can be confusing.  

https://docs.docker.com/engine/reference/builder/#expose

File: main.go (hunk 2 above)
Comment (justincely) [reply main.go:11]: Good, yeah so when you bind to `127.0.0.1` that means you'll reject anything that does not come from the same host (more or less, don't worry about it) and if you bind to `0.0.0.0` it'll accept from anywhere.  and if you set it to `244.10.8.24` it'll only accept connections from that specific host.

So we want it to default to 127.0.0.1 so that if we mistakenly forget to set it it'll be secure.  But now that we ran into the issue of needing it to be multiple values, it's time to make this configurable at runtime.  Either via environment variables or a config file (or both) we need to be able to set this when we run it. 

- https://github.com/spf13/cobra
- https://github.com/spf13/viper

**Comment by ZachSellman** [comment]
Alright here's a quick attempt at what I think you were asking for. I'm a little fuzzy on the exact specifics around this regarding how configuration should be taking place ideally. Docker build and other commands allow you to specify various config details prior to building, is that similar to what we should be going for here?
I guess this boils down to: Is it reasonable to expect others who use a project you built to just open the config file and manually change it or is there some other way this should be done?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+FROM golang:1.19 AS build-stage
+RUN mkdir /app
+WORKDIR /app
+COPY . .
+RUN go mod download
+RUN CGO_ENABLED=0 GOOS=linux go build -o main ./...
+
+
+FROM alpine:latest AS production
+COPY --from=build-stage /app .
```
[the remark this answers is not in the document]
Comment (justincely) [reply Dockerfile:12]: @ZachSellman

**Review by justincely — COMMENTED** [review]
wow forgot to push the button

File: Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (justincely) [reply Dockerfile:12]: one minor thing, just to understand it.  Then we can merge this and move on to the python.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+RUN mkdir /app
+WORKDIR /app
+COPY . .
+RUN go mod download
+RUN CGO_ENABLED=0 GOOS=linux go build -o main ./...
+
+
+FROM alpine:latest AS production
+COPY --from=build-stage /app/main .
+COPY config.yaml .
```
Comment (justincely) [inline Dockerfile:13]:
```suggestion
COPY --from=build-stage /app/config.yaml .
```

Comment (justincely) [reply Dockerfile:13]: technically the same, but still.  Basically move the source code in just once, then copy it to wherever it needs to be.

**Review by justincely — APPROVED** [review]

File: Dockerfile (hunk 4 above)
Comment (ZachSellman) [reply Dockerfile:13]: oh duh lol

**Review by justincely — APPROVED** [review]
